src/bsp/platform/posix/xi_bsp_io_fs_posix.c

```diff
@@ -0,0 +1,328 @@
+/* Copyright (c) 2003-2017, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xi_bsp_io_fs.h>
+#include <xi_bsp_mem.h>
+
+#include <xi_list.h>
+
+#include <stdio.h>
+#include <sys/stat.h>
+#include <errno.h>
+#include <assert.h>
+#include <memory.h>
+
+/* The size of the buffer to be used for reads. */
+const size_t xi_fs_buffer_size = 1024;
+
+#define XI_BSP_IO_FS_CHECK_CND( cnd, e, s )                                              \
+    if ( ( cnd ) )                                                                       \
+    {                                                                                    \
+        ( s ) = ( e );                                                                   \
+        goto err_handling;                                                               \
+    }
+
+/**
+ * @brief xi_fs_posix_file_handle_container_t
+ *
+ * Database type of file handles and allocated memory chunks.
+ */
+typedef struct xi_fs_posix_file_handle_container_s
+{
+    FILE* posix_fp;
+    uint8_t* memory_buffer;
+    struct xi_fs_posix_file_handle_container_s* __next;
+} xi_fs_posix_file_handle_container_t;
+
+/* local list of open files */
+static xi_fs_posix_file_handle_container_t* xi_fs_posix_files_container;
+
+/* helper function that translates the errno errors to the xi_state_t values */
+static xi_state_t xi_fs_posix_errno_2_xi_state( int errno_value )
+{
+    xi_state_t ret = XI_STATE_OK;
+
+    switch ( errno_value )
+    {
+        case EBADF:
+        case EACCES:
+        case EFAULT:
+        case ENOENT:
+        case ENOTDIR:
+            ret = XI_FS_RESOURCE_NOT_AVAILABLE;
+            break;
+        case ELOOP:
+        case EOVERFLOW:
+            ret = XI_FS_ERROR;
+            break;
+        case ENAMETOOLONG:
+            ret = XI_INVALID_PARAMETER;
+            break;
+        case ENOMEM:
+            ret = XI_OUT_OF_MEMORY;
+            break;
+        default:
+            ret = XI_FS_ERROR;
+    }
+
+    return ret;
+}
+
+/* helper function that translates posix stat to xi stat */
+static xi_state_t xi_fs_posix_stat_2_xi_stat( const struct stat* const posix_stat,
+                                              xi_fs_stat_t* const xi_stat )
+{
+    assert( NULL != posix_stat );
+    assert( NULL != xi_stat );
+
+    xi_stat->resource_size = posix_stat->st_size;
+
+    return XI_STATE_OK;
+}
+
+/**
+ * @brief xi_fs_posix_file_list_cnd
+ *
+ * Functor for list find function.
+ *
+ * @param list
+ * @param fp
+ * @return 1 if list element is the one with the matching fp 0 otherwise
+ */
+static uint8_t
+xi_fs_posix_file_list_cnd( xi_fs_posix_file_handle_container_t* list, FILE* fp )
+{
+    assert( NULL != list );
+    assert( NULL != fp );
+
+    return ( list->posix_fp == fp ) ? 1 : 0;
```

Comment:
This function is a functor - naming is from C++. Used in lines 200, 256 and 290 by XI_LIST_FIND macros. The FIND macros uses this function to check each element whether it is the searched one. Actually this is like a "bool std::operator==( const A& op1, const A& op2);" function in C++.
And yes, I think the list parameter should be renamed to list_element, I'll do this.